eZ/Publish/Core/Repository/URLService.php

```diff
@@ -0,0 +1,232 @@
+        }
+
+        if ($query->limit !== null && !is_numeric($query->limit)) {
+            throw new InvalidArgumentValue('limit', $query->limit);
+        }
+
+        $results = $this->urlHandler->find($query);
+
+        $items = [];
+        foreach ($results['items'] as $url) {
+            $items[] = $this->buildDomainObject($url);
+        }
+
+        return new SearchResult([
+            'count' => $results['count'],
+            'items' => $items,
+        ]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function updateUrl(URL $url, URLUpdateStruct $struct)
+    {
+        if ($this->repository->hasAccess('url', 'update') !== true) {
+            throw new UnauthorizedException('url', 'update');
+        }
+
+        if (!$this->isUnique($url->id, $struct->url)) {
+            throw new InvalidArgumentException('struct', 'url already exists');
+        }
+
+        $updateStruct = $this->buildUpdateStruct($this->loadById($url->id), $struct);
+
+        $this->repository->beginTransaction();
+        try {
+            $this->urlHandler->updateUrl($url->id, $updateStruct);
+            $this->repository->commit();
+        } catch (\Exception $e) {
+            $this->repository->rollback();
+            throw $e;
+        }
+
+        return $this->loadById($url->id);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadById($id)
+    {
+        if ($this->repository->hasAccess('url', 'view') !== true) {
+            throw new UnauthorizedException('url', 'view');
+        }
+
+        return $this->buildDomainObject(
+            $this->urlHandler->loadById($id)
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadByUrl($url)
+    {
+        if ($this->repository->hasAccess('url', 'view') !== true) {
+            throw new UnauthorizedException('url', 'view');
+        }
+
+        return $this->buildDomainObject(
+            $this->urlHandler->loadByUrl($url)
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createUpdateStruct()
+    {
+        return new URLUpdateStruct();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function findUsages(URL $url, $offset = 0, $limit = -1)
+    {
+        $usages = $this->urlHandler->findUsages($url->id);
+
+        $query = new Query();
+        if (!empty($usages)) {
+            $query->filter = new ContentCriterion\LogicalAnd([
+                new ContentCriterion\ContentId($usages),
+                new ContentCriterion\Visibility(ContentCriterion\Visibility::VISIBLE),
+            ]);
+        } else {
+            $query->filter = new ContentCriterion\MatchNone();
+        }
+
+        $query->offset = $offset;
+        if ($limit > -1) {
+            $query->limit = $limit;
+        }
+
+        return $this->repository->getSearchService()->findContentInfo($query);
+    }
+
+    protected function buildDomainObject(SPIUrl $data)
```

Comment:
Note: Usually this should be a responsibility of `DomainMapper`, but at that point I'm unsure myself because it's just one small method ;)

However please add PhpDoc.

Reply:
Done in 0faac18c2e88bcf94b44f3452104d681c3ea9ecb.